tests/components/nest/test_api.py

```diff
@@ -0,0 +1,106 @@
+"""Tests for the Nest integration API glue library."""
+
+import time
+
+from homeassistant.components.nest.api import AsyncConfigEntryAuth
+from homeassistant.components.nest.const import (
+    OAUTH2_AUTHORIZE,
+    OAUTH2_TOKEN,
+    SDM_SCOPES,
+)
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.util import dt
+
+from tests.common import MockConfigEntry
+
+DOMAIN = "nest"
+
+CLIENT_ID = "some-client-id"
+CLIENT_SECRET = "some-client-secret"
+
+FAKE_TOKEN = "some-token"
+FAKE_UPDATED_TOKEN = "some-updated-token"
+FAKE_REFRESH_TOKEN = "some-refresh-token"
+
+EXPIRATION_TIME = time.time() + 86400
+
+CONFIG_ENTRY_DATA = {
+    "sdm": {},  # Indicates new SDM API, not legacy API
+    "auth_implementation": "local",
+    "token": {
+        "access_token": FAKE_TOKEN,
+        "expires_in": 12354,  # ignored
+        "refresh_token": FAKE_REFRESH_TOKEN,
+        "scope": " ".join(SDM_SCOPES),
+        "token_type": "Bearer",
+        "expires_at": EXPIRATION_TIME,
+    },
+}
+
+
+def create_auth(hass, aioclient_mock, config_entry_data):
+    """Create an AsyncConfigEntryAuth for tests."""
+    config_entry = MockConfigEntry(domain=DOMAIN, data=CONFIG_ENTRY_DATA)
+    config_entry.add_to_hass(hass)
+    implementation = config_entry_oauth2_flow.LocalOAuth2Implementation(
+        hass, DOMAIN, CLIENT_ID, CLIENT_SECRET, OAUTH2_AUTHORIZE, OAUTH2_TOKEN
+    )
+    session = config_entry_oauth2_flow.OAuth2Session(hass, config_entry, implementation)
+    return AsyncConfigEntryAuth(aioclient_mock, session, CLIENT_ID, CLIENT_SECRET)
+
+
+async def test_get_access_token(hass, aioclient_mock):
+    """Verify that the access token is loaded from the ConfigEntry."""
+    auth = create_auth(hass, aioclient_mock, CONFIG_ENTRY_DATA)
+
+    token = await auth.async_get_access_token()
+    assert token == FAKE_TOKEN
+
+
+async def test_get_creds(hass, aioclient_mock):
+    """Verify that the Credentials are created properly."""
+    auth = create_auth(hass, aioclient_mock, CONFIG_ENTRY_DATA)
+
+    creds = await auth.async_get_creds()
```

Comment:
We want to patch the library so we can assert the return value of `async_get_creds` without patching `async_get_creds` and without calling `async_get_creds` directly.

Reply:
For subscriber based tests.  Now using a `FakeSubscriberFactory` to capture the creds, which I think partially addresses this comment.  However, I'm not sure I agree this is the right test approach and there is an inconsistency with testing between the api vs subscriber.  Let me give some more context/background and we can see if there is an even better approach or future interface/interfaces I should improve since the AbstractAuth pattern may be failing.

The [AbstractAuth](https://github.com/allenporter/python-google-nest-sdm/blob/master/google_nest_sdm/auth.py) pattern is based on the recommended approach, but has a couple tweaks since it is used by two APIs:
- `request`: Public method to call the API
- `async_get_access_token`: Implemented by child class, used internally by `AbstractAuth`
- `async_get_creds: Used by the `GoogleNestSubscriber` when creating a subscriber.  At the moment this is considered part of the "public" API of `AbstractAuth`
- `AbstractSubscriberFactory` / `DefaultSubscriberFactory`: Used by `GoogleNestSubscriber` to actually create the subscriber, and it gets `creds` passed in

Question #1: For API based tests, I did not yet update these to change to a similar approach.   I'm taking your comment to mean that similarly there we shouldn't be testing `async_get_access_token` directly, but instead testing based on the `request` method?

Question #2: Do we need to improve the interface /design for subscriber auth? Currently `async_get_creds` seemed like the right level because its the "public" API for how the [GoogleNestSubscriber](https://github.com/allenporter/python-google-nest-sdm/blob/349619450e1fa4ac2de8ae0be282b0778e8365b0/google_nest_sdm/google_nest_subscriber.py#L132) gets the auth, differing from how the API based auth works.
- Maybe `AbstractAuth` is doing too much.  This may be case where it should prefer composition rather than inheritance (pull out auth creation into a standalone class not mixed in with request logic, breaking from the Home Assistant pattern).  This is keeping with he status quo of sharing auth logic between the API and susbcriber cases.
- Maybe the equivalent design would be to add an `async_get_creds` method to `AbstractSubscriberFactory` and callers implement that.  This would be a larger change to split out auth separately for API and Subscriber.
Do you have thoughts about this? Curious if you know what I mean about the public interfaces.

Reply:
I'm not sure about the library design. I think you know that best. As long as we're using the Home Assistant API correctly and things work well I'm happy.

Regarding tests, I've tried to answer that above.

Reply:
OK, all this was based on my total misunderstanding of your comment, thinking we were talking about "unit testing".